src/nemos/solvers.py

```diff
@@ -0,0 +1,726 @@
+                # ),
+            )
+
+            return OptStep(params=xs, state=state)
+
+        # at the end of each epoch, check for convergence or reaching the max number of epochs
+        def cond_fun(step):
+            _, state = step
+            return (state.iter_num <= self.maxiter) & (state.error >= self.tol)
+
+        final_xs, final_state = loop.while_loop(
+            cond_fun=cond_fun,
+            body_fun=body_fun,
+            init_val=OptStep(params=init_params, state=init_state),
+            maxiter=self.maxiter,
+            jit=True,
+        )
+        return OptStep(params=final_xs, state=final_state)
+
+    @partial(jit, static_argnums=(0,))
+    def _update_per_random_samples(
+        self,
+        current_params: Any,
+        state: SVRGState,
+        prox_lambda: float,
+        *args,
+    ) -> OptStep:
+        """
+        Performs the inner loop of Prox-SVRG sweeping through approximately one full epoch,
+        updating the parameters after sampling a mini-batch on each iteration.
+
+        Parameters
+        ----------
+        current_params : Any
+            Parameters at the end of the previous update, used as the starting point for the current update.
+        state : SVRGState
+            Optimizer state at the end of the previous sweep.
+            Needs to have the current anchor point (xs) and the gradient at the anchor point (df_xs) already set.
+        prox_lambda : float
+            Regularization strength. Can be None.
+        args:
+            Positional arguments passed to loss function `fun` and its gradient.
+            For GLMs it is:
+                X :
+                    Input data.
+                y :
+                    Output data.
+
+        Returns
+        -------
+        OptStep
+            xs : Any
+                Parameters at the end of the last inner loop.
+                (... or the average of the parameters over the last inner loop)
+            state : SVRGState
+                Updated state.
+        """
+        n_points_per_arg = {leaf.shape[0] for leaf in jax.tree.leaves(args)}
+        if not len(n_points_per_arg) == 1:
+            raise ValueError("All arguments must have the same sized first dimension.")
+        N = n_points_per_arg.pop()
+
+        m = (N + self.batch_size - 1) // self.batch_size  # number of iterations
+        # m = N
+
+        xs, df_xs = state.xs, state.df_xs
+
+        def inner_loop_body(_, carry):
+            xk, x_sum, key = carry
+
+            # sample mini-batch or data point
+            key, subkey = random.split(key)
+            ind = random.randint(subkey, (self.batch_size,), 0, N)
+
+            # perform a single update on the mini-batch or data point
+            xk = self._xk_update(
+                xk,
+                xs,
+                df_xs,
+                state.stepsize,
+                prox_lambda,
+                *(tree_slice(arg, ind) for arg in args),
```

Comment:
```suggestion
                *tree_slice(args, ind),
```

This is sufficient because you are sure that args is a tuple, the output of a tree_map is going to be a tuple (same tree struct as args), and you can unpack output of tree slice directly.

Reply:
Cool!